src/event/sys/unix/parse.rs

```diff
@@ -311,32 +412,54 @@ pub(crate) fn parse_csi_u_encoded_key_code(buffer: &[u8]) -> Result<Option<Inter
             (KeyModifiers::NONE, KeyEventKind::Press)
         };
 
-    let keycode = {
-        if let Some(c) = char::from_u32(codepoint) {
-            match c {
-                '\x1B' => KeyCode::Esc,
-                '\r' => KeyCode::Enter,
-                // Issue #371: \n = 0xA, which is also the keycode for Ctrl+J. The only reason we get
-                // newlines as input is because the terminal converts \r into \n for us. When we
-                // enter raw mode, we disable that, so \n no longer has any meaning - it's better to
-                // use Ctrl+J. Waiting to handle it here means it gets picked up later
-                '\n' if !crate::terminal::sys::is_raw_mode_enabled() => KeyCode::Enter,
-                '\t' => {
-                    if modifiers.contains(KeyModifiers::SHIFT) {
-                        KeyCode::BackTab
-                    } else {
-                        KeyCode::Tab
+    let (keycode, state) = {
+        if let Some((special_key_code, state)) = translate_functional_key_code(codepoint) {
+            (special_key_code, state)
+        } else if let Some(c) = char::from_u32(codepoint) {
+            (
+                match c {
+                    '\x1B' => KeyCode::Esc,
+                    '\r' => KeyCode::Enter,
+                    // Issue #371: \n = 0xA, which is also the keycode for Ctrl+J. The only reason we get
+                    // newlines as input is because the terminal converts \r into \n for us. When we
+                    // enter raw mode, we disable that, so \n no longer has any meaning - it's better to
+                    // use Ctrl+J. Waiting to handle it here means it gets picked up later
+                    '\n' if !crate::terminal::sys::is_raw_mode_enabled() => KeyCode::Enter,
+                    '\t' => {
+                        if modifiers.contains(KeyModifiers::SHIFT) {
+                            KeyCode::BackTab
+                        } else {
+                            KeyCode::Tab
+                        }
                     }
-                }
-                '\x7F' => KeyCode::Backspace,
-                _ => KeyCode::Char(c),
-            }
+                    '\x7F' => KeyCode::Backspace,
+                    _ => KeyCode::Char(c),
+                },
+                KeyEventState::empty(),
+            )
         } else {
             return Err(could_not_parse_event_error());
         }
     };
 
-    let input_event = Event::Key(KeyEvent::new_with_kind(keycode, modifiers, kind));
+    if let KeyCode::Modifier(modifier_keycode) = keycode {
+        match modifier_keycode {
+            ModifierKeyCode::LeftAlt | ModifierKeyCode::RightAlt => {
+                modifiers.set(KeyModifiers::ALT, true)
+            }
+            ModifierKeyCode::LeftControl | ModifierKeyCode::RightControl => {
+                modifiers.set(KeyModifiers::CONTROL, true)
+            }
+            ModifierKeyCode::LeftShift | ModifierKeyCode::RightShift => {
```

Comment:
I am not sure what IsoLevel3Shift, IsoLevel5Shift are. But should we consider adding those here?

Reply:
I don't think so; those are extra modifiers used in keyboard layouts like [Neo](https://en.wikipedia.org/wiki/Neo_(keyboard_layout)) to switch to extra punctuation/number layers.